### ISSUE TITLE
New SOTA Method (FSQ)

### ISSUE THREAD

**Issue by mahdip72** [lucidrains/vector-quantize-pytorch#239]
Hello,

Thanks for this amazing work!

Apparently this is the new SOTA VQ model that achived 100% codebook utilization and lowest error rate.

[Scalable Training for Vector-Quantized Networks with 100% Codebook Utilization](https://arxiv.org/abs/2509.10140v1)

Any plan on adding this to your amazing library?

**Comment by lucidrains** [comment]
@mahdip72 you know that's a huge red flag right? lol

**Comment by lucidrains** [comment]
@mahdip72 i could throw the modification into a branch and you can try it out? i just don't want to add stuff that doesn't end up working as advertised

**Comment by mahdip72** [comment]
@lucidrains Yeah I will. I can test it quickly with my recent protein structure tokenizer: https://github.com/mahdip72/vq_encoder_decoder

I have used different quantization types (learnable, residual and FSQ) before for that project and all of them works reasonably well. If that is fine, I would use the same project for testing.

**Comment by mahdip72** [comment]
> <img alt="Image" width="240" src="https://private-user-images.githubusercontent.com/108653/531343091-38e4f8ab-ccbc-4077-a60f-f2e08520dc64.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjczMTkzMzQsIm5iZiI6MTc2NzMxOTAzNCwicGF0aCI6Ii8xMDg2NTMvNTMxMzQzMDkxLTM4ZTRmOGFiLWNjYmMtNDA3Ny1hNjBmLWYyZTA4NTIwZGM2NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEwMlQwMTU3MTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02N2I1NjczYzBiZjk0N2QwMzQwMDFiNzIyMWE3YWI2NDUzZDhjODkyN2Y3MjA1MWMwM2ZmY2JiNzkzYjkyZWI3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.66GqzaJqHHnb7aMGoR7S_ZsBk-rRg4aCmmOlVYeuYDo"> took this somewhere in SF. thought it was funny how the acronym takes on a whole new meaning

Seems you found where the baselines live!


> [@mahdip72](https://github.com/mahdip72) at the project root, do `git checkout test-fvq`, then run `python autoencoder_fvq.py` for fashionmnist. you can also do `pip install .` to then try it for your protein autoencoder

Appreciate it.

**Comment by mahdip72** [comment]
Hi Phil @lucidrains 
Sorry for the late response. I have been drained by the recent incidents in my country, Iran. 

Yeah, the HPs are finicky. I did some experiments for this FVQ feature on my protein stuff project. At first, it did not work well, even for a common vocab size (4096). But after tweaking the vq_bridge transformer architecture, I could make it work, at least for my project. I am trying to find the most critical hyperparameters for it.

For now, I can reach 98% valid-set utilization with a 16k codebook and, similar to the official implementation ([this](https://github.com/yfChang-cv/FVQ)), a four-block transformer stack. And it turns out that including the positional encoding and reducing the patch size (combining embeddings before giving them to the vq-bridge) play an important role. 

I will update here if I find more insights on it.

**Comment by lucidrains** [comment]
> > [@mahdip72](https://github.com/mahdip72) thank you Mahdi! that is very helpful
> > is that x-axis in training steps? how do your final reconstruction values look?
> 
> X-axis is the number of epochs (each has 20k samples). The second figure is like the final metric I am monitoring for the project. The `vq bridge 256 dim, patch size 1` was better than the vanilla one.

perfect thanks!

> > is that a finding on your end, or did the authors note this too?
> 
> I did not see the authors mentioning that. My intuition was that the role of vq bridge is not necessarily to move information between codebooks in the transformer, but rather to stabilize the gradients for codebooks learning. So, I guess that register token is like the attention sink token of LLMs to avoid over-mixing.

yes, very possible

> > however, out of curiosity, i ran an experiment with just a single attention + register + no patching, and it did the worse reconstruction wise (while code util looks good). i then ablated the attention but kept the single feedforward, and that one performed the best. could you run these experiments on your end and let me know what you find?
> 
> Yeah sure.

don't worry about this, as both runs came short of having both, so your configuration was still optimal!

> > this makes me want to dig up an old paper where they applied this to transformer "code" embeddings (this was in the DEQ / ALBERT era, if you recall those papers). i'll share it if i can find it again as it may provide some clue to what's going on
> 
> I was a junior ML engineer when ALBERT came out! That feels like a century ago now. Indeed, many genuine ideas are published early on, but nobody bothers to revisit them and do extensive ablations to see what actually works.

yes me too, the days when it took me a month to fully digest a paper (and had to attend all these paper reading clubs)
